src/OpenTelemetry/Metrics/MetricReader.cs

```diff
@@ -163,27 +137,60 @@ internal virtual void SetParentProvider(BaseProvider parentProvider)
             this.ParentProvider = parentProvider;
         }
 
-        protected abstract bool OnCollect(Batch<Metric> metrics, int timeoutMilliseconds);
+        /// <summary>
+        /// Processes a batch of metrics.
+        /// </summary>
+        /// <param name="metrics">Batch of metrics to be processed.</param>
+        /// <param name="timeoutMilliseconds">
+        /// The number of milliseconds to wait, or <c>Timeout.Infinite</c> to
+        /// wait indefinitely.
+        /// </param>
+        /// <returns>
+        /// Returns <c>true</c> when metrics processing succeeded; otherwise,
+        /// <c>false</c>.
+        /// </returns>
+        protected abstract bool ProcessMetrics(Batch<Metric> metrics, int timeoutMilliseconds);
 
         /// <summary>
-        /// Called by <c>ForceFlush</c>. This function should block the current
-        /// thread until flush completed, shutdown signaled or timed out.
+        /// Called by <c>Collect</c>. This function should block the current
+        /// thread until metrics collection completed, shutdown signaled or
+        /// timed out.
         /// </summary>
         /// <param name="timeoutMilliseconds">
         /// The number of milliseconds to wait, or <c>Timeout.Infinite</c> to
         /// wait indefinitely.
         /// </param>
         /// <returns>
-        /// Returns <c>true</c> when flush succeeded; otherwise, <c>false</c>.
+        /// Returns <c>true</c> when metrics collection succeeded; otherwise,
+        /// <c>false</c>.
         /// </returns>
         /// <remarks>
         /// This function is called synchronously on the thread which called
-        /// <c>ForceFlush</c>. This function should be thread-safe, and should
+        /// <c>Collect</c>. This function should be thread-safe, and should
         /// not throw exceptions.
         /// </remarks>
-        protected virtual bool OnForceFlush(int timeoutMilliseconds)
+        protected virtual bool OnCollect(int timeoutMilliseconds)
         {
-            return true;
+            var sw = Stopwatch.StartNew();
+
+            var collectMetric = this.ParentProvider.GetMetricCollect();
+            var metrics = collectMetric();
+
+            if (timeoutMilliseconds == Timeout.Infinite)
+            {
+                return this.ProcessMetrics(metrics, Timeout.Infinite);
+            }
+            else
+            {
+                var timeout = timeoutMilliseconds - sw.ElapsedMilliseconds;
+
+                if (timeout <= 0)
+                {
+                    return false;
+                }
+
+                return this.ProcessMetrics(metrics, (int)timeout);
```

Comment:
If `var timeout = timeoutMilliseconds - sw.ElapsedMilliseconds;` asigns `0.1ms` then it is rounded down to `0` here: `(int)timeout`. 

I think it is OK as b/c later I read:

```csharp
// notify all the readers, even if we run overtime
result = cur.Value.Collect((int)Math.Max(timeout, 0)) && result;
```

but it would be nice to give a comment at least like:

```csharp
return this.ProcessMetrics(metrics, (int)timeout); // ProcessMetrics supports timeout=0
```

I am proposing it, only because I spend like 10 min analyzing if it is OK to pass `0` to `ProcessMetrics`. `if (timeout <= 0)` was suggesting that it is not.

Reply:
I've updated the XML comments to call out what is a valid timeout value. PTAL.

Reply:
Resolve please